storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -428,6 +452,16 @@ private void deleteFinalizedBlocks(final List<Pair<UInt64, Bytes32>> blocksToPru
             pair -> updater.deleteFinalizedBlock(pair.getLeft(), pair.getRight()));
         updater.commit();
       }
+      // retrieve earliest finalized slot left after pruning and set it in the DB
+      final Optional<UInt64> earliestBlockSlot =
+          dao.getEarliestFinalizedBlock().map(SignedBeaconBlock::getSlot);
+
+      if (earliestBlockSlot.isPresent()) {
+        try (final FinalizedUpdater updater = finalizedUpdater()) {
+          updater.setEarliestBlockSlot(earliestBlockSlot.get());
+          updater.commit();
+        }
+      }
```

Comment:
Actually this kind of screams database inconsistency, which was my concern with storing this in a database. We might need to think about this a little more. If we're storing this variable in the database at all, it needs to be maintained inside the same tx importing earlier blocks, and in the same tx pruning... otherwise we're not going to be able to trust the value at any point in time...

I think inside the context here, we just need to check before we commit the deletion... the earliest block fn will be inside the tx context at that point...

Reply:
Agree that we should be updating that inside the same tx to ensure the integrity of the value we're storing but changes don't get effectively made in the DB until we commit them. I could move this fetch/set to inside the first   try(updater) statement but that to me feels like holding the tx from being committed unnecessarily. 

On the other hand I don't see how we would update this concurrently. My understanding is that we are always effectively executing one or the other(downloading historic blocks or pruning them), never both at the same time. Is that correct?

Reply:
I have done some changes to address the issue you spotted and some others I found during the tests